willofsteel/async_client.py

```diff
@@ -0,0 +1,333 @@
+"""
+WillofSteel API Wrapper
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Asynchronous wrapper for the Will of Steel API.
+
+:copyright: (C) 2024-present ItsNeil
+:license: MIT, see LICENSE for more details
+
+"""
+from __future__ import annotations
+from typing import Literal
+import logging
+
+import aiohttp
+import asyncio
+
+from .types import Player, Alliance, MarketOrder, UnitType, ItemType, LoggingObject, convert_str_to_IT, convert_str_to_UT, Outpost
+from .constants import BASE, ALL_ITEMS, MISSING
+from .utils import parse_error, setup_logging
+from .exceptions import *
+
+
+class AsyncClient:
+    """
+    Asynchronous client for the Will of Steel API.
+
+    Parameters
+    ----------
+    api_key: :class:`str`
+        The API key to use.
+    logger: :class:`~willofsteel.types.LoggingObject`
+        The logger to use. Defaults to MISSING.
+    use_existing_loop: :class:`bool`
+        Whether to use an existing event loop. Defaults to False.
+
+    """
+
+    def __init__(self, api_key: str, logger: LoggingObject = MISSING):
+        self.api_key = api_key
+        self.headers = {
+            "API-Key": self.api_key,
+            "User-Agent": "Will of Steel API Wrapper",
+            "Accept": "application/json",
+            "X-API-Version": "0.3"
+        }
+        self._session = None
+        setup_logging(logger if logger else LoggingObject())
+
+        event_loop = asyncio.get_event_loop()
+        if event_loop.is_running():
+            event_loop.create_task(self._async_setup())
+        else:
+            asyncio.run(self._async_setup())
+
+    async def _async_setup(self):
+        """
+        Setup the asynchronous parts of the client.
+
+        """
+        self._session = aiohttp.ClientSession()
+        await self._verify_key()
+
+    async def close(self):
+        """
+        Close the AsyncClient when it no longer needs to be used.
+
+        """
+        await self._session.close()
+
+    async def _verify_key(self) -> None:
+        data, status = await self.request("GET", "/verify", self.headers)
+        if status == 403:
+            raise InvalidKey
+        elif status == 200:
+            logging.info("Key verification successful.")
+        else:
+            raise ServerError
+
```

Comment:
Add helper methods such as `get_user_id` (Returns the user ID returned by the API on the verification call), `is_verified` (can be used just incase, returns bool).

These would just be QoL methods for any developers if they need them.